libp2p/network/connection/swarm_connection.py

```diff
@@ -43,6 +43,9 @@ def __init__(self, muxed_conn: IMuxedConn, swarm: "Swarm") -> None:
         # We *could* optimize this but it really isn't worth it.
         for stream in self.streams:
             await stream.reset()
+        # Force context switch for stream handlers to process the stream reset event we just emit
+        # before we cancel the stream handler tasks.
+        await asyncio.sleep(0.1)
```

Comment:
This is a [code smell](https://ethpm-cli.readthedocs.io/en/latest/)

I would suggest looking into a mechanism to allow you to directly `await` this type of thing.  For example, `await stream.reset()` could return an object that could then be awaited and wouldn't return until the underlying handlers have processed the event.  Otherwise you're likely to continue to have to letter random sleeps into your code and I can say from experience in trinity that we don't want this.

Reply:
`await stream.reset()` along with `await self.muxed_conn.close()` on L40 will `set` the `event`s for `Mplex` and `Mplex_stream`. This signals that `Mplex` and its Mplex_stream`s are shutting down. Then when stream handler try to read from stream it will catch `StreamReset` error and properly finishes.

So the problem is that changing the state of the stream does not guarantee that the stream handler task that's reading from the stream will detect the state change and finish right after. So in this case if we cancel tasks right after `await stream.reset()` the stream handler task will be cancelled instead of catching `StreamReset` error and properly finishes.

Maybe we can cancel other tasks and wait for stream handler task to finish?